content/blog/2022-05-04-e2e-cv-pt2.md

```diff
@@ -0,0 +1,197 @@
+author: alex_kim
+commentsUrl: https://discuss.dvc.org/t/end-to-end-computer-vision/1178
+tags:
+  - Computer Vision
+  - DVC
+  - CML
+  - Studio
+  - CI/CD
+  - Experiment Tracking
+---
+
+### Introduction
+
+Earlier, we built a pipeline that produces a trained Computer Vision model. Now
+we need a way to efficiently tune its configuration and the hyperparameters of
+the model. We want the ability to:
+
+- Run many experiments and easily compare their results to pick the
+  best-performing ones.
+- Track the global history of the model's performance, and map each improvement
+  to a particular change in code, configuration, or data.
+- Zoom into the details of each training run to help us diagnose issues.
+
+### Experiment Management
+
+Our DVC pipeline relies on the parameters defined in the
+[`params.yaml`](https://github.com/iterative/magnetic-tiles-defect/blob/main/params.yaml)
+file. By loading its contents in each stage, we can avoid hard-coded parameters.
+It also allows rerunning the whole or parts of our pipeline under a different
+set of parameters. The DVC pipeline YAML file
+[`dvc.yaml`](https://github.com/iterative/magnetic-tiles-defect/blob/main/dvc.yaml)
+supports a
+[templating format](https://dvc.org/doc/user-guide/project-structure/pipelines-files#templating)
+to insert values from different sources in the YAML structure itself.
+
+DVC tracks which stages of the pipeline experienced changes and only reruns
+those. By changes, we mean _everything_ that might affect the predictive
+performance of your model like changes to the dataset, source code and/or
+parameters. This not only ensures complete reproducibility but often
+significantly reduces the time needed to rerun the whole pipeline while ensuring
+consistent results on every rerun. For example, if we only change the code in
+the evaluation stage (e.g., change what type of metrics we want to compute),
+there is no reason for us to rerun the often time-consuming data preprocessing
+and model training stages:
+
+```bash
+$ dvc exp run
+Running stage 'check_packages':
+> pipenv run pip freeze > requirements.txt
+Stage 'data_load' didn't change, skipping
+Stage 'data_split' didn't change, skipping
+Stage 'train' didn't change, skipping
+Running stage 'evaluate':
+> python src/stages/eval.py --config=params.yaml
+...
+```
+
+There is a super convenient set of
+[Experiment Management](https://dvc.org/doc/user-guide/experiment-management)
+features that make switching between reproducible experiments very easy without
+adding failed experiments to your git history. Check out this
+[blog post](https://dvc.org/blog/ml-experiment-versioning), which talks about
+the idea of "ML Experiments as Code." That means treating experiments as you'd
+treat code, that is, use git to track all changes in configs, metrics, and data
+versions through text files. This approach removes the need for a separate
+database/online service to store experiment metadata. If we run a few
+experiments:
```

Comment:
It seems like there's something missing after "If we run a few experiments"? Then what?

Reply:
yup, fixed